ext/storage_sources/local_store/local_store.c

```diff
@@ -729,58 +739,65 @@ local_open(WT_FILE_SYSTEM *file_system, WT_SESSION *session, const char *name,
     LOCAL_FILE_SYSTEM *local_fs;
     LOCAL_FLUSH_ITEM *flush;
     LOCAL_STORAGE *local;
-    WT_FILE_HANDLE *file_handle;
-    int fd, oflags, ret;
-    char *open_name;
+    WT_FILE_HANDLE *file_handle, *wt_fh;
+    WT_FILE_SYSTEM *wt_fs;
+    struct stat sb;
+    int ret;
+    bool exists;
 
     (void)flags; /* Unused */
 
-    fd = oflags = ret = 0;
+    ret = 0;
     *file_handlep = NULL;
     local_fh = NULL;
     local_fs = (LOCAL_FILE_SYSTEM *)file_system;
     local = local_fs->local_storage;
+    wt_fs = local_fs->wt_fs;
 
-    if (file_type != WT_FS_OPEN_FILE_TYPE_DATA)
-        return (
-          local_err(local, session, EINVAL, "%s: open: only data file types supported", name));
-
-    local->op_count++;
-    if (flags == WT_FS_OPEN_CREATE)
-        oflags = O_WRONLY | O_CREAT;
-    else if (flags == WT_FS_OPEN_READONLY)
-        oflags = O_RDONLY;
-    else
-        return (local_err(local, session, EINVAL, "open: invalid flags: 0x%x", flags));
+    /*
+     * We expect that the local file system will be used narrowly, like when creating or opening a
+     * data file or turtle file. It would be unexpected to try to open a non-data file (like a log
+     * file) in that narrow part of code, so we make it an error here.
+     *
+     * Relaxing this constraint to allow opening of, say, log files, would be straightforward - we
+     * would not translate the path or do any tracking for flushing. But there's a catch. Other
+     * parts of the API, like remove and rename, have no flag indicating that they are operating on
+     * a log file, so we wouldn't know whether to do path translation. Of course, we could peek at
+     * the name, but that would be bad form.
+     */
+    if (file_type != WT_FS_OPEN_FILE_TYPE_DATA && file_type != WT_FS_OPEN_FILE_TYPE_REGULAR)
+        return (local_err(
+          local, session, EINVAL, "%s: open: only data file and regular types supported", name));
 
     /* Create a new handle. */
     if ((local_fh = calloc(1, sizeof(LOCAL_FILE_HANDLE))) == NULL) {
         ret = ENOMEM;
         goto err;
     }
-    if ((ret = local_location_path(file_system, name, NULL, &local_fh->path)) != 0)
+    if ((ret = local_location_path(file_system, name, &local_fh->path)) != 0)
         goto err;
-    if (flags == WT_FS_OPEN_CREATE) {
+
+    ret = stat(local_fh->path, &sb);
+    if (ret != 0 && errno != ENOENT)
+        ret = local_err(local, session, errno, "%s: local_open stat", local_fh->path);
+    exists = (ret == 0);
+
+    /*
+     * TODO: tiered: If the file doesn't exist locally, make a copy of it from the cloud here.
```

Comment:
Sanity checking my thinking here. 

If the file is in the cloud and we don't have a local copy, could we just read the blocks we need from cloud, without doing the copy?

Even if we eventually want to copy the object to the local file system, it might be nice to handle an immediate read of a few KB without waiting to download a multi-MB (or multi-GB) object first.

I'm just thinking out loud. No need to change anything here.

Reply:
Yes, we could.  I've not (yet) implemented any removal from the local directory, but when we do, we could explore what you're talking about it.  The easiest solution is to get the whole file every time of course.  But in fact what we'll probably have a client of this fs (i.e. the block manager) that has a file handle still open, and it eventually will do a read of a block at an offset.  What we don't know is whether we'll have more reads coming, but the right thing may well be to "page in" just the one block or a chunk around it.  It would require some per-file tracking, but totally doable.